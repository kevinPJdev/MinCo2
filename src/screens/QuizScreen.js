import {
  TouchableOpacity,
  StyleSheet,
  Text,
  View,
  FlatList,
  Alert,
} from "react-native";
import React from "react";
import { Colors } from "../style/colors";
import QuizButtonScreen from "../components/Button/QuizButtonScreen";
import { useNavigation } from "@react-navigation/native";

const styles = StyleSheet.create({
  pageHeading: {
    color: "#FFF",
    textAlign: "center",
  },
  container: {
    flex: 1,
    backgroundColor: Colors.green20,
    flexDirection: "column",
    paddingTop: 20,
  },
  scrollContainer: {
    padding: 10,
  },
  question: {
    padding: 10,
    borderColor: Colors.green10,
    borderWidth: 3,
    borderRadius: 10,
    margin: 5,
  },
  questionTitle: {
    marginBottom: 6,
  },
  optionContainer: {
    flex: 1,
    borderColor: Colors.green10,
    borderWidth: 3,
    backgroundColor: Colors.green40,
    borderRadius: 10,
    paddingVertical: 5,
    flexDirection: "column",
    paddingHorizontal: 10,
    marginBottom: 10,
    alignItems: "baseline",
  },
  option: {
    color: "#FFF",
  },
  selectedOption: {
    color: "#FFF",
    textAlign: "center",
    opacity: 0.5,
  },
});

/**
 * Screen for showing Quiz to user
 * as per the data stored locally
 */
const QuizScreen = ({ data }) => {
  const [selectedOptions, setSelectedOptions] = React.useState({});
  /**
   * Don't allow submit until all questions's answers are selected
   */
  const isSubmitDisabled =
    Object.keys(selectedOptions).length != (data || []).length;
  /**
   * Hook to manage navigation
   */
  const navigation = useNavigation();

  /**
   * Responsible for updating state
   * as per option selected
   */
  const selectAnswer = (questionIndex, answer) => {
    setSelectedOptions({
      ...selectedOptions,
      [questionIndex]: answer,
    });
  };

  /**
   * Renders each question
   */
  const questionMapper = ({ item, index }) => {
    const [question, questionIndex] = [item, index];
    return (
      <View style={styles.question}>
        <Text style={styles.questionTitle}>{question.text || ""} </Text>
        {(question.options || []).map((option) => {
          const isOptionSelected =
            selectedOptions[questionIndex] == option.value;
          return (
            <View key={option.label}>
              <TouchableOpacity
                activeOpacity={0.8}
                onPress={() => selectAnswer(questionIndex, option.value)}
                style={styles.optionContainer}
              >
                <Text
                  style={
                    isOptionSelected ? styles.selectedOption : styles.option
                  }
                >
                  {option.label}
                </Text>
              </TouchableOpacity>
            </View>
          );
        })}
      </View>
    );
  };

  const showQuizResult = () => {
    /**
     * Calculate Total emission from the selected options
     */
    const totalEmission = Object.keys(selectedOptions)
      .map((questionIndex) => selectedOptions[questionIndex])
      .reduce((total, current) => (total += current), 0);
    /**
     * Show result in a Dialog
     * and redirect to Emission screen
     */
    Alert.alert("Quiz Result", `Your Co2 emission is ${totalEmission} kgCO2`, [
      {
        text: "Thanks!",
        onPress: () => navigation.navigate("Budget"),
      },
    ]);
  };

  return (
    <View style={styles.container}>
      <Text style={styles.pageHeading}>WELCOME TO QUIZ SCREEN</Text>

      <FlatList
        data={data || []}
        renderItem={questionMapper}
        style={styles.scrollContainer}
      />
      <QuizButtonScreen
        title="Submit Quiz"
        onPress={showQuizResult}
        disabled={isSubmitDisabled}
      />
    </View>
  );
};

export default QuizScreen;
