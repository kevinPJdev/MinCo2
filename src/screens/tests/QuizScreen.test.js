import React from "react";
import renderer from "react-test-renderer";
import QuizScreen from "../QuizScreen";

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: () => {},
    }),
  };
});

const testData = [
  {
    text: "Testing.",
    options: [
      {
        value: 0.2604,
        label: "Testing",
      },
    ],
  },
];

describe("Render Quiz Screen", () => {
  it("With Null data", () => {
    const tree = renderer.create(<QuizScreen />).toJSON();
    expect(tree.children.length).toBe(3);
  });

  it("With Proper data", () => {
    const tree = renderer.create(<QuizScreen data={testData} />).toJSON();
    expect(tree.children.length).toBe(3);
  });

  it("Test snapshot", () => {
    const tree = renderer.create(<QuizScreen data={testData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
