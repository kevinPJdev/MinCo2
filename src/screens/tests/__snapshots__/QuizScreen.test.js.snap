// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Quiz Screen Test snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#66BDAB",
      "flex": 1,
      "flexDirection": "column",
      "paddingTop": 20,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFF",
        "textAlign": "center",
      }
    }
  >
    WELCOME TO QUIZ SCREEN
  </Text>
  <RCTScrollView
    data={
      Array [
        Object {
          "options": Array [
            Object {
              "label": "Testing",
              "value": 0.2604,
            },
          ],
          "text": "Testing.",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "padding": 10,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderColor": "#A8C8C1",
              "borderRadius": 10,
              "borderWidth": 3,
              "margin": 5,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Testing.
             
          </Text>
          <View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "baseline",
                  "backgroundColor": "#164239",
                  "borderColor": "#A8C8C1",
                  "borderRadius": 10,
                  "borderWidth": 3,
                  "flex": 1,
                  "flexDirection": "column",
                  "marginBottom": 10,
                  "opacity": 1,
                  "paddingHorizontal": 10,
                  "paddingVertical": 5,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#FFF",
                  }
                }
              >
                Testing
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#164239",
        "borderRadius": 10,
        "elevation": 8,
        "marginLeft": 140,
        "marginRight": 50,
        "opacity": 1,
        "paddingHorizontal": 12,
        "paddingVertical": 10,
        "width": "30%",
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#FFFFFF",
          "fontSize": 13,
          "fontWeight": "bold",
          "textTransform": "uppercase",
        }
      }
    >
      Submit Quiz
    </Text>
  </View>
</View>
`;

exports[`Render Quiz Screen Test snapshot data 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#66BDAB",
      "flex": 1,
      "flexDirection": "column",
      "paddingTop": 20,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFF",
        "textAlign": "center",
      }
    }
  >
    WELCOME TO QUIZ SCREEN
  </Text>
  <RCTScrollView
    data={
      Array [
        Object {
          "options": Array [
            Object {
              "label": "Testing",
              "value": 0.2604,
            },
          ],
          "text": "Testing.",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "padding": 10,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderColor": "#A8C8C1",
              "borderRadius": 10,
              "borderWidth": 3,
              "margin": 5,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 6,
              }
            }
          >
            Testing.
             
          </Text>
          <View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "baseline",
                  "backgroundColor": "#164239",
                  "borderColor": "#A8C8C1",
                  "borderRadius": 10,
                  "borderWidth": 3,
                  "flex": 1,
                  "flexDirection": "column",
                  "marginBottom": 10,
                  "opacity": 1,
                  "paddingHorizontal": 10,
                  "paddingVertical": 5,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#FFF",
                  }
                }
              >
                Testing
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#164239",
        "borderRadius": 10,
        "elevation": 8,
        "marginLeft": 140,
        "marginRight": 50,
        "opacity": 1,
        "paddingHorizontal": 12,
        "paddingVertical": 10,
        "width": "30%",
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#FFFFFF",
          "fontSize": 13,
          "fontWeight": "bold",
          "textTransform": "uppercase",
        }
      }
    >
      Submit Quiz
    </Text>
  </View>
</View>
`;
